### DISCUSSION TITLE
New "filterIf" method on model service

### DISCUSSION THREAD

**Discussion by stephengalbraith** [ExpressionEngine/ExpressionEngine#3216] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
The model service was introduced some time ago and while a major leap forward at the time for EE, I think today it lags behind some powerhouse frameworks (e.g., Laravel, Symfony).

At the moment when it comes to conditionally filtering on the model service we are required to use `if` statements (or even `switch` sometimes).

For example:

```PHP
$entry = ee('Model')->get('ChannelEntry');

if($filter['status'] == 'show_only_closed'){
    $entry->filter('status', 'closed');
}

$result = $entry->all();
```

It would be nice to be able to do one consistent chain on the Model builder.

**Describe the solution you'd like**

Taken from [Laravel's Conditionable Trait](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Support/Traits/Conditionable.php#L15), they use the following:

```PHP
public function when($value, $callback, $default = null)
{
    if ($value) {
        return $callback($this, $value) ?: $this;
    } elseif ($default) {
        return $default($this, $value) ?: $this;
    }

    return $this;
}
```

Sticking with EE's current naming style, we could do something like this (and it need not even be chained necessarily):

```PHP
$entry = ee('Model')->get('ChannelEntry')
    ->filterIf($filter['status'] == 'show_only_closed', function($builder){
        return $builder->filter('status', 'closed');
    })
    ->all();
```

This is also just one of many usages of this method.

**Describe alternatives you've considered**
Beyond sticking with a litany of if/switch statements, I haven't considered any other alternatives, but do regularly use the Laravel framework and its ORM.

**Teachability, Documentation, Adoption, Migration Strategy**
If there seems to be any appetite for this I wouldn't be opposed to writing up something for the docs.

- [x] I am capable and would like to work on implementation of this feature if it is considered.
- [x] This would also be backwards-compatible as well given it doesn't yet exist in the EE codebase and wouldn't be in use anywhere.

**Comment by intoeetive** [comment]
@stephengalbraith I guess my first question would be, does it solve any problem, apart from allowing to chain the code better?

@TomJaeger @bryannielsen calling for your Laravel expertise here
